### PR TITLE
workflow: move yaml parsing to server side

### PR DESCRIPTION
**Pull request by masco-enovance**

### CODE REVIEW

**Comment by safchain** [comment]
I would propose something like this

https://paste.fedoraproject.org/paste/x4zPoEnbMWPMcd9CYjD4EA

@lebauce @masco-enovance what do you think ?

**Comment by masco-enovance** [comment]
@safchain I changed the code as you suggested

**Comment by masco-enovance** [comment]
run skydive-compile-tests
run skydive-functional-tests-backend-elasticsearch
run skydive-k8s-tests
run skydive-scale-tests

**Comment by masco-enovance** [comment]
run skydive-cdd-overview-tests
run skydive-functional-tests-backend-orientdb
run skydive-k8s-tests
run skydive-scale-tests

File: cmd/client/alert.go [hunk 1]
```diff
@@ -71,7 +71,7 @@ var AlertCreate = &cobra.Command{
 			exitOnError(err)
 		}
 
-		if err := client.Create("alert", &alert); err != nil {
+		if err := client.Create("alert", &alert, "json"); err != nil {
```
Comment (nplanel) [inline cmd/client/alert.go]: may define "json"/"yaml" as a const type in common/types.go  (a kind like SortOrder)
type APIMessageFormat int

Comment (masco-enovance) [reply cmd/client/alert.go]: @nplanel Done, Thanks!

File: http/client.go [hunk 2]
```diff
@@ -149,14 +152,24 @@ func (c *CrudClient) Get(resource string, id string, value interface{}) error {
...
-	s, err := json.Marshal(value)
-	if err != nil {
-		return err
+func (c *CrudClient) Create(resource string, value interface{}, contentType common.APIMessageFormat, ret ...interface{}) error {
+	var s []byte
+	var err error
+	var header = http.Header{}
+	switch contentType {
+	case common.YAML:
+		s = []byte(value.(string))
```
Comment (lebauce) [inline http/client.go]: What about using the same mechanism that for JSON and use `yaml.Marshal` ?

File: cmd/client/workflow.go [hunk 3]
```diff
@@ -89,12 +78,13 @@ var WorkflowCreate = &cobra.Command{
...
 
-		workflow, err := loadWorklow(workflowPath)
+		yml, err := loadWorklow(workflowPath)
 		if err != nil {
 			exitOnError(err)
 		}
 
-		if err := client.Create("workflow", &workflow); err != nil {
+		var workflow types.Workflow
+		if err := client.Create("workflow", string(yml), common.YAML, &workflow); err != nil {
```
Comment (lebauce) [inline cmd/client/workflow.go]: I would probably be more consistent to pass the workflow as an object, and not as a string

File: api/server/server.go [hunk 4]
```diff
@@ -133,7 +136,18 @@ func (a *Server) RegisterAPIHandler(handler Handler, authBackend shttp.Authentic
 
 				resource := handler.New()
 
-				if err := common.JSONDecode(r.Body, &resource); err != nil {
+				var err error
+				if contentType := r.Header.Get("Content-Type"); contentType == "application/yaml" {
+					if content, e := ioutil.ReadAll(r.Body); e == nil {
+						err = yaml.Unmarshal(content, resource)
```
Comment (lebauce) [inline api/server/server.go]: For the `yaml.Unmarshal` to work, we need to add `yaml:MyAttribute` tags for all the API resources (right now, only the `Workflow` type have these tags)

**Comment by masco-enovance** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by masco-enovance** [comment]
run skydive-functional-tests-backend-elasticsearch
run skydive-functional-tests-backend-orientdb

**Comment by masco-enovance** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by masco-enovance** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by masco-enovance** [comment]
rerun skydive-functional-tests-backend-orientdb

**Comment by masco-enovance** [comment]
rerun skydive-functional-tests-backend-orientdb

**Comment by masco-enovance** [comment]
run skydive-functional-tests-backend-orientdb

File: api/types/types.go [hunk 5]
```diff
@@ -70,22 +70,22 @@ func NewAlert() *Alert {
...
-	LayerKeyMode   string           `json:"LayerKeyMode,omitempty" valid:"isValidLayerKeyMode"`
-	ExtraLayers    flow.ExtraLayers `json:"ExtraLayers,omitempty"`
+	BasicResource  `yaml:",inline"`
+	GremlinQuery   string           `json:"GremlinQuery,omitempty" valid:"isGremlinExpr" yaml:"GremlinQuery"`
+	BPFFilter      string           `json:"BPFFilter,omitempty" valid:"isBPFFilter" yaml:"BPFFilter"`
+	Name           string           `json:"Name,omitempty" yaml:"Name"`
+	Description    string           `json:"Description,omitempty" yaml:"Description"`
+	Type           string           `json:"Type,omitempty" yaml:"Type"`
+	Count          int              `json:"Count" yaml:"Count"`
+	PCAPSocket     string           `json:"PCAPSocket,omitempty" yaml:"BPFFilter"`
```
Comment (lebauce) [inline api/types/types.go]: s/BPFFilter/PACAPSocket/

File: api/types/types.go [hunk 6]
```diff
@@ -142,25 +142,25 @@ func (n *NodeRule) Validate() error {
...
+	Dst              string `yaml:"Dst"`
+	SrcIP            string `yaml:"SrcIP"`
+	DstIP            string `yaml:"DstIP"`
+	SrcMAC           string `yaml:"SrcMAC"`
+	DstMAC           string `yaml:"DstMAC"`
+	SrcPort          int64  `yaml:"SrcPort"`
+	DstPort          int64  `yaml:"DstPort"`
+	Type             string `yaml:"Type"`
+	Payload          string `yaml:"Payload"`
 	TrackingID       string
```
Comment (lebauce) [inline api/types/types.go]: Any reason to not put a tag for TrackingID ?

Comment (masco-enovance) [reply api/types/types.go]: It is generated from server side, so i skipped a tag

File: api/types/types.go (hunk 5 above)
Comment (masco-enovance) [reply api/types/types.go]: good catch, copy cat error ;)

**Comment by masco-enovance** [comment]
run skydive-functional-tests-backend-elasticsearch
run skydive-functional-tests-backend-orientdb

**Comment by masco-enovance** [comment]
run skydive-functional-tests-backend-orientdb

**Pull request merged**